### ISSUE TITLE
Transitioning from picking tree nodes to picking resources

### ISSUE THREAD

**Issue by alexweininger** [microsoft/vscode-azureresourcegroups#521]
**tl;dr**: Apart from v1.5 compatibility related code, v2 clients (and resource groups) should stop using a quick pick resource implementation that relies on a tree data provider. This prevents breaking changes as a result of changing the tree UX, and allows us to stop exposing the tree data providers on the v2 API. It also improves code clarity within commands. My proposed solution is to just have quick pick steps list resources directly from their source (Azure SDK, HTTP request, etc.). This solution fits nicely into another pattern we'll adopt when rewriting extensions for v2, which is to stop attaching state to tree nodes and instead attach it to the resource a tree node represents (see #515).

"quick pick resource" = picking any resource via a series of quick picks

We still need to have a v1.5 compatibility quick pick resource implementation that uses the tree data provider as long as we have v1.5 clients to support. And I also think it's fine to provide a shared implementation of a quick pick resource implementation that uses a tree data provider, as long as we have a consuming extension (Docker maybe?). However, in the case of the Azure Resources extension, and integrating client extensions, I strongly believe we should not use such implementation.

I'll present why v2 clients and RGs should stop using an quick pick resource implementation that is based on a tree data provider below.

#### 1. Changing the structure of the tree causes breaking changes

Making sure all the quick pick resource usages are still supported has accounted for a large portion of the work in both the release of v1, and again during the release of v2 (implementing the v1.5 compatibility shim).

I'd love to say that we'll never change the tree view UX again, but at the same time I know we're going to change it. Although it's in a good place now, the unified UX is still relatively new and could need changes at any time.

Also, this reliance on the tree UX hampers our flexibility in the resources extension. Once we have v2 clients relying on the layout of the tree, it'll be an even larger pain to change.

#### 2. We no longer cache tree nodes, which was one of the main reasons quick pick resource was implemented using the tree data provider

Before the v1 release, tree nodes were cached, and the quick pick resource implementation utilized this cache to provide a snappy experience when picking any nodes that had already been exposed in the tree.

#### 3. Picked tree nodes will never be the same as tree nodes that are displayed on the tree (because no cache)

As I outlined in https://github.com/microsoft/vscode-azureresourcegroups/issues/487#issuecomment-1376616009, a tree data provider based quick pick resource implementation returns tree nodes that are confusingly disconnected from the tree view.

This can easily cause bugs as it's easy to forget that the returned `AppSettingItem` is not equal to the `AppSettingItem` on the tree, even if they represent the same app setting. We should be picking the resource the nodes represent, and not the nodes themselves.

#### 4. Remove the tree data providers from the v2 Resources API, reducing the API surface

If we stop using a tree data provider based quick pick resource implementation, there's no reason to continue exposing the Azure and Workspace tree data providers on the v2 API. I think this is really important as the smaller we can keep our API, the less we need to maintain and support.

### Proposed solution

Moving forward, I think we should use a quick pick resource implementation that is based on listed the resources directly. For example, listing subscriptions from the Azure Account API, listing Container Apps from the SDK, app settings from SDK, and so on.

We can easily add common quick pick steps to our shared packages

We can easily implement shared utilities for the common steps like pick subscription. And we can create shared base classes for a generic pick resource step.

#### Benefits

Apart from the reasoning I gave above, one of the largest benefits is eliminating the magic box nature of the current pattern

Quick pick resource logic will be contained within the command using it, eliminating the magic box nature of the current pattern. It should be clear what prompts are shown when a command is invoked via the command palette. Currently this is not the case as the prompts are determined by logic scattered across `AzExtTreeItem`, the tree nodes (`pickTreeItemImpl`, `isAncestorOf`), and the calling command.

This solution will force implementors to follow the correct patterns as it concerns updating tree items. There won't be any confusion surrounding "Are these tree nodes from the picker, or from the tree?".

It'll help enforce that any state needs to be mapped to the resource a tree node represents, rather than the tree node itself.

#### Drawbacks (and rebuttals 😄)

The code required for quick pick resource in each command will be greater than with the current pattern. However, this is for the sake of clarity and simplicity. And I argue that this tradeoff is totally worth it.

I know we won't be able to just do `pickResource(contextValue)`, however, as I pointed out above, I never loved that solution as it's never clear which prompts will be shown. For simple commands that act on tree nodes near the top of the tree it's fine, but it quickly gets convoluted for the deeper tree nodes (skipping UI only nodes, excluding nodes, customizing quick pick labels, etc).

When we had each developer assigned to a single extension this wasn't as big of an issue. Since it was easier to remember the details about each command and the prompt steps used to quick pick the particular resource needed for that command. And remember the context values of each item.

However, that isn't the case anymore. And each time a developer has to debug a quick pick resource usage, it involves a deeper dive than what I prefer for something that should be simple.

Instead of `pickResource(contextValue)`, we can do the following (in psuedocode)

```
run wizard with these steps:
	PickSubscriptionStep
	PickFunctionAppStep
	PickAppSettingStep
```

A majority of our commands require just 2 or 3 steps, like the one above. The longest quick pick resource we have would be for the Open GitHub Action Log command in Static Web Apps, which has 6 steps including picking subscription.

```
run wizard with these steps:
	PickSubscriptionStep
	PickStaticWebAppStep
	PickEnvironmentStep
	PickActionStep
	PickJobStep
	PickStepStep
```

And I think that's fine since it's an outlier, plus I'm willing to trade this verbosity for clarity. In that I'm able to see clearly what quick pick resource is doing. And like I've mentioned, say we redesign the Static Web App tree items, with the proposed pattern, we don't have to worry about breaking the quick pick resource code. Also, the code for listing each resource can easily be shared between quick pick steps and tree nodes.

**Comment by nturinski** [comment]
> 1. Changing the structure of the tree causes breaking changes

This actually is already kind of a problem because I always felt it was weird to prompt for the "group type" nodes depending on the tree's grouping. Plus, it'll give us a lot of flexibility to implement things like favorites or pinning without impacting the picker in anyway.

**Comment by bwateratmsft** [comment]
This all sounds reasonable to me. The key advantage of the current solution is convenience--it's easy to deal with commands that can be performed through either tree item context menus or through the command palette, as they can quickly join into a single code path (usually in the first lines of our commands where we check if `node` is falsy, and do a pick if so).

I would say that for any use case in which we are not using the quick picks _strictly as a substitute for tree navigation_, we should not be using the TDPs at all, and should do as you're suggesting--get the data "through your own means". Using the quick picks more rigidly as an alternate mode of tree navigation precludes create buttons and potentially step skipping.

A key disadvantage of the above, that I don't see mentioned, is that to acquire information about resources that "belong" to another extension means either establishing a backchannel for them to share information or including another copy of the SDK. It would be logistically very inconvenient to try to keep those packages in sync across multiple extensions, so the net result would be more memory usage and longer load times. Consequently, I'd prefer the "backchannel" approach, but _how_ that would be done needs to be approached very carefully and thoughtfully.

@alexweininger and I discussed a potential alternate/supplemental/something solution, wherein we put one more layer inbetween what we give to VSCode and what we're getting from the primary TDPs / BDPs. We could use this layer to do some extra caching as appropriate, and do things like `runWithTemporaryDescription`, track deleted nodes, etc. That being said, I don't really think those two cases offer enough value to justify the complexity they demand.
